Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -9,6 +16,346 @@ pub use num::NumCast;
+    }
+
+    fn to_array(&self) -> Rc<Vec<T>> {
+        // We convert the match above to statements using the Rust "it" idiom
+        if let Sequence::ArraySequence { values, .. } = self {
+            // The length of the elements
+            Rc::clone(values)
+        } else if let Sequence::ConcatSequence { length, .. } = self {
+            let mut array: Vec<T> = Vec::with_capacity(*length);
+            Sequence::<T>::append_recursive(&mut array, self);
+            Rc::new(array)
+        } else if let Sequence::LazySequence { boxed, is_string, ..  } = self {
+            let result = boxed.borrow().to_array();
+            // Put the value back into boxed
+            boxed.replace(Rc::clone(
+                &Sequence::<T>::new_array_sequence_is_string(&result, *is_string)));
+            result
+        } else {
+            panic!("This should never happen")
+        }
+    }
+
+    fn append_recursive(array: &mut Vec<T>, this: &Sequence<T>) {
+        match this {
+            Sequence::ArraySequence { values, .. } =>
+              // The length of the elements
+              for value in values.iter() {
+                array.push(value.clone());
+              },
+            Sequence::ConcatSequence { left, right, .. } =>
+              // Let's create an array of size length and fill it up recursively
+              {
+                Sequence::<T>::append_recursive(array, left);
+                Sequence::<T>::append_recursive(array, right);
+              }
+            Sequence::LazySequence { boxed, .. } =>
+            Sequence::<T>::append_recursive(array, boxed.borrow().as_ref()),
+        }
+    }
+
+    fn node_count(&self) -> usize {
+        match self {
+            Sequence::ArraySequence { node_count, .. } =>
+              // The length of the elements
+              *node_count,
+            Sequence::ConcatSequence { node_count, ..  } =>
+              *node_count,
+            Sequence::LazySequence { node_count, .. } =>
+              *node_count,
+        }
+    
+    }
+    /// Returns the cardinality of this [`Sequence<T>`].
+    // The cardinality returns the length of the sequence
+    fn cardinality(&self) -> SizeT {
+        match self {
+            Sequence::ArraySequence { values, .. } =>
+              // The length of the elements
+              values.len(),
+            Sequence::ConcatSequence { length, .. } =>
+              *length,
+            Sequence::LazySequence { length, .. } =>
+              *length,
+        }
+    }
+    fn select(&self, index: SizeT) -> T {
+        let array = self.to_array();
+        array[index].clone()
+    }
+    
+}
+
+// **************
+// Immutable maps
+// **************
+
+#[allow(dead_code)]
+struct Map<K, V> 
+  where K: Clone + Eq + std::hash::Hash, V: Clone
+{
+    data: Rc<Sequence<(K, V)>>,
+    // Any time we explicitly access this map, we index the data
+    cache: RefCell<Option<HashMap<K, V>>>
+}
+
+#[allow(dead_code)]
+impl <K, V> Map<K, V>
```

Comment:
Ah now this is interesting. On the one hand great, ALL backends should have a smarter implementation of immutable maps. On the other hand the code we want to support with this backend has already had to work around the lack of this in the existing backends, so this optimization is not going to help them - they will use https://github.com/dafny-lang/libraries/tree/master/src/MutableMap instead.

I'd say don't spend time on this optimization here, go with the simple wrapping of a mutable map like the other backends. At some point we should do what you're doing here, but in `dafnyRuntime.dfy` instead so that we have a shot at getting consistent runtime performance across all backends.

Reply:
Since I already completed this optimization, let's keep it for now as it is strictly better in terms of complexity than wrapping a hashmap. Wrapping a hashmap is O(n) when you add a key/value and O(1) when accessing it, whereas this is O(1) when adding a key/value pair or even combining maps, and amortized O(1) when accessing it.

But either optimizations are not even the panacea, as if there is a loop where someone tests if something is in the map before adding it, then it defeats totally the optimization as the Hashmap will be realized at every step anyway.

My vision is that for those who really want performance, they will use real mutable hashmap for which maps will only be ghost specification.
In other terms, I believe that classes and arrays should be the most optimized, whereas built-in immutable types are mostly there for prototyping and can tolerate non-optimal complexities.

Reply:
That's all reasonable. Let me refine my advice:

This does look like a nice optimization without making the implementation a lot more complex. But it is still slightly more complex. In particular, you have to worry about `data` and `cache` getting out of sync when used by different concurrent executions. That problem is solved for the lazy sequences in `dafnyRuntime.dfy` by using an `AtomicBox`, and for Rust it looks like the equivalent solution is a `RefCell`, so perhaps you need that here.

All I'm saying is that you still have more work TODO to complete this implementation anyway, so I'm still not convinced that this optimization is worth the non-zero risk. I think it would be more worth it if multi-targeting Dafny code could rely on those [amortized] O(1) costs regardless of the backend.

Reply:
Concurrency is a good thing to consider, but consider this:
For this code to work under concurrent execution, we would need anyway to use the slower Arc instead of Rc, because Rc is not concurrent-friendly. And I'm not sure we want to so that, we still need to evaluate that need.

But regarding your remark, how can data and cache get ouf of sync? I'm already using RefCell
Data is only ever written once during construction, and cache could theoretically be updated concurrently but all result only depend only on the data so at worst it will be populated twice with the same value.